ddtrace/internal/products.py

```diff
@@ -0,0 +1,178 @@
+from ddtrace.internal.uwsgi import uWSGIMasterProcess
+
+
+log = get_logger(__name__)
+
+if sys.version_info < (3, 10):
+    from importlib_metadata import entry_points
+else:
+    from importlib.metadata import entry_points
+
+try:
+    from typing import Protocol  # noqa:F401
+except ImportError:
+    from typing_extensions import Protocol  # type: ignore[assignment]
+
+
+class Product(Protocol):
+    requires: t.List[str]
+
+    def post_preload(self) -> None:
+        ...
+
+    def start(self) -> None:
+        ...
+
+    def restart(self, join: bool = False) -> None:
+        ...
+
+    def stop(self, join: bool = False) -> None:
+        ...
+
+    def at_exit(self, join: bool = False) -> None:
+        ...
+
+
+class ProductManager:
+    __products__: t.Dict[str, Product] = {}  # All discovered products
+
+    def __init__(self) -> None:
+        self._products: t.Optional[t.List[t.Tuple[str, Product]]] = None  # Topologically sorted products
+
+        for product_plugin in entry_points(group="ddtrace.products"):
+            name = product_plugin.name
+            log.debug("Discovered product plugin '%s'", name)
+
+            # Load the product protocol object
+            try:
+                product: Product = product_plugin.load()
+            except Exception:
+                log.exception("Failed to load product plugin '%s'", name)
+                continue
+
+            log.debug("Product plugin '%s' loaded successfully", name)
+
+            self.__products__[name] = product
+
+    def _sort_products(self) -> t.List[t.Tuple[str, Product]]:
+        # Data structures for topological sorting
+        q: t.Deque[str] = deque()  # Queue of products with no dependencies
+        g = defaultdict(list)  # Graph of dependencies
+        f = {}  # Remaining dependencies for each product
+
+        for name, product in self.__products__.items():
+            product_requires = getattr(product, "requires", [])
+            if not product_requires:
+                q.append(name)
+            else:
+                f[name] = list(product_requires)
+                for r in product_requires:
+                    g[r].append(name)
+
+        # Determine the product (topological) ordering
+        ordering = []
+        while q:
+            n = q.popleft()
+            ordering.append(n)
+            for p in g[n]:
+                f[p].remove(n)
+                if not f[p]:
+                    q.append(p)
+                    del f[p]
+
+        if f:
+            log.error(
+                "Circular dependencies among products detected. These products won't be enabled: %s.", list(f.keys())
+            )
+
+        return [(name, self.__products__[name]) for name in ordering if name not in f]
+
+    @property
+    def products(self) -> t.List[t.Tuple[str, Product]]:
+        if self._products is None:
+            self._products = self._sort_products()
+        return self._products
+
+    def start_products(self) -> None:
```

Comment:
Generally speaking, if products are being sorted based on dependency/topological order, does that mean earlier products failing to start up will cause problems for later products?

Reply:
Quite likely yes! We should probably store the outcome of the operation and prevent dependant products from starting (with some informative logging).